File: aether_messenger/windows/AetherQMessenger/Crypto/Identity.cs
using System;
using System.IO;
using System.Security.Cryptography;
using System.Text.Json;
using AetherQMessenger.Ffi;

namespace AetherQMessenger.Crypto;

/// <summary>
/// Qurilmaning o'z uzoq-muddatli identity kalitlar to'plami: Ed25519 (imzo), X25519
/// (statik DH), ML-KEM-768 (post-kvant KEM) — AUTH-AKEM handshake'ida ishlatiladi.
/// <c>ZRecip</c> — Implicit Rejection uchun <c>XSk</c> bilan birga saqlanadigan alohida
/// maxfiy qiymat (spec 4.1-bo'lim, band 4).
/// </summary>
public sealed class Identity
{
    public required string UserId { get; init; }
    public required byte[] EdPk { get; init; }
    public required byte[] EdSk { get; init; }
    public required byte[] XPk { get; init; }
    public required byte[] XSk { get; init; }
    public required byte[] MlkemPk { get; init; }
    public required byte[] MlkemSk { get; init; }
    public required byte[] ZRecip { get; init; }
}

internal sealed class IdentityFile
{
    public string UserId { get; set; } = "";
    public string EdPk { get; set; } = "";
    public string EdSkProtected { get; set; } = "";
    public string XPk { get; set; } = "";
    public string XSkProtected { get; set; } = "";
    public string MlkemPk { get; set; } = "";
    public string MlkemSkProtected { get; set; } = "";
    public string ZRecipProtected { get; set; } = "";
}

/// <summary>
/// Identity — Windows DPAPI (<see cref="ProtectedData"/>, CurrentUser scope) bilan
/// himoyalangan mahalliy faylda saqlanadi. Ochiq kalitlar aniq matnda, maxfiy kalitlar
/// esa har biri alohida DPAPI-shifrlangan holda yoziladi.
/// </summary>
public static class IdentityStore
{
    private static string FilePath =>
        Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
            "AetherQMessenger",
            "identity.json");

    public static Identity GetOrCreate()
    {
        if (File.Exists(FilePath))
        {
            var json = File.ReadAllText(FilePath);
            var file = JsonSerializer.Deserialize<IdentityFile>(json)
                       ?? throw new InvalidOperationException("identity.json buzilgan yoki bo'sh");

            return new Identity
            {
                UserId = file.UserId,
                EdPk = Convert.FromBase64String(file.EdPk),
                EdSk = Unprotect(file.EdSkProtected),
                XPk = Convert.FromBase64String(file.XPk),
                XSk = Unprotect(file.XSkProtected),
                MlkemPk = Convert.FromBase64String(file.MlkemPk),
                MlkemSk = Unprotect(file.MlkemSkProtected),
                ZRecip = Unprotect(file.ZRecipProtected),
            };
        }

        var edKp = AetherCrypto.GenerateEd25519Identity();
        var xKp = AetherCrypto.GenerateX25519Identity();
        var mlkemKp = AetherCrypto.GenerateMlKem768Identity();
        var zRecip = AetherCrypto.RandomBytes(32);

        var userIdBytes = SHA256.HashData(edKp.PublicKey)[..8];
        var userId = Convert.ToHexString(userIdBytes).ToLowerInvariant();

        var identity = new Identity
        {
            UserId = userId,
            EdPk = edKp.PublicKey,
            EdSk = edKp.SecretKey,
            XPk = xKp.PublicKey,
            XSk = xKp.SecretKey,
            MlkemPk = mlkemKp.PublicKey,
            MlkemSk = mlkemKp.SecretKey,
            ZRecip = zRecip,
        };

        Save(identity);
        return identity;
    }

    private static void Save(Identity identity)
    {
        var dir = Path.GetDirectoryName(FilePath)!;
        Directory.CreateDirectory(dir);

        var file = new IdentityFile
        {
            UserId = identity.UserId,
            EdPk = Convert.ToBase64String(identity.EdPk),
            EdSkProtected = Protect(identity.EdSk),
            XPk = Convert.ToBase64String(identity.XPk),
            XSkProtected = Protect(identity.XSk),
            MlkemPk = Convert.ToBase64String(identity.MlkemPk),
            MlkemSkProtected = Protect(identity.MlkemSk),
            ZRecipProtected = Protect(identity.ZRecip),
        };

        File.WriteAllText(FilePath, JsonSerializer.Serialize(file));
    }

    private static string Protect(byte[] data) =>
        Convert.ToBase64String(ProtectedData.Protect(data, null, DataProtectionScope.CurrentUser));

    private static byte[] Unprotect(string base64) =>
        ProtectedData.Unprotect(Convert.FromBase64String(base64), null, DataProtectionScope.CurrentUser);
}

File: aether_messenger/android/app/src/main/java/com/aetherq/messenger/crypto/Identity.kt
package com.aetherq.messenger.crypto

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import java.security.MessageDigest

/**
 * Qurilmaning o'z uzoq-muddatli identity kalitlar to'plami: Ed25519 (imzo), X25519
 * (statik DH), ML-KEM-768 (post-kvant KEM) — AUTH-AKEM handshake'ida ishlatiladi.
 * `zRecip` — Implicit Rejection uchun `sk_x25519` bilan birga saqlanadigan alohida
 * maxfiy qiymat (spec 4.1-bo'lim, band 4).
 */
data class Identity(
    val userId: String,
    val edPk: ByteArray,
    val edSk: ByteArray,
    val xPk: ByteArray,
    val xSk: ByteArray,
    val mlkemPk: ByteArray,
    val mlkemSk: ByteArray,
    val zRecip: ByteArray,
)

/** Identity — Android Keystore bilan himoyalangan `EncryptedSharedPreferences`da saqlanadi. */
class IdentityStore(context: Context) {

    private val prefs: SharedPreferences = run {
        val masterKey = MasterKey.Builder(context).setKeyScheme(MasterKey.KeyScheme.AES256_GCM).build()
        EncryptedSharedPreferences.create(
            context,
            "aetherq_identity",
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
        )
    }

    /** Mavjud identity'ni qaytaradi, bo'lmasa yangisini generatsiya qilib saqlaydi. */
    fun getOrCreate(): Identity {
        prefs.getString(KEY_USER_ID, null)?.let { userId ->
            return Identity(
                userId = userId,
                edPk = hexToBytes(prefs.getString(KEY_ED_PK, "")!!),
                edSk = hexToBytes(prefs.getString(KEY_ED_SK, "")!!),
                xPk = hexToBytes(prefs.getString(KEY_X_PK, "")!!),
                xSk = hexToBytes(prefs.getString(KEY_X_SK, "")!!),
                mlkemPk = hexToBytes(prefs.getString(KEY_MLKEM_PK, "")!!),
                mlkemSk = hexToBytes(prefs.getString(KEY_MLKEM_SK, "")!!),
                zRecip = hexToBytes(prefs.getString(KEY_Z_RECIP, "")!!),
            )
        }

        val edKp = AetherCrypto.generateEd25519Identity()
        val xKp = AetherCrypto.generateX25519Identity()
        val mlkemKp = AetherCrypto.generateMlKem768Identity()
        val zRecip = AetherCrypto.randomBytes(32)
        val userId = MessageDigest.getInstance("SHA-256").digest(edKp.publicKey).take(8).toByteArray().toHex()

        prefs.edit()
            .putString(KEY_USER_ID, userId)
            .putString(KEY_ED_PK, edKp.publicKey.toHex())
            .putString(KEY_ED_SK, edKp.secretKey.toHex())
            .putString(KEY_X_PK, xKp.publicKey.toHex())
            .putString(KEY_X_SK, xKp.secretKey.toHex())
            .putString(KEY_MLKEM_PK, mlkemKp.publicKey.toHex())
            .putString(KEY_MLKEM_SK, mlkemKp.secretKey.toHex())
            .putString(KEY_Z_RECIP, zRecip.toHex())
            .apply()

        return Identity(userId, edKp.publicKey, edKp.secretKey, xKp.publicKey, xKp.secretKey, mlkemKp.publicKey, mlkemKp.secretKey, zRecip)
    }

    companion object {
        private const val KEY_USER_ID = "user_id"
        private const val KEY_ED_PK = "ed_pk"
        private const val KEY_ED_SK = "ed_sk"
        private const val KEY_X_PK = "x_pk"
        private const val KEY_X_SK = "x_sk"
        private const val KEY_MLKEM_PK = "mlkem_pk"
        private const val KEY_MLKEM_SK = "mlkem_sk"
        private const val KEY_Z_RECIP = "z_recip"
    }
}
